daemon/logger/awslogs/cloudwatchlogs.go

```diff
@@ -309,48 +344,83 @@ var newTicker = func(freq time.Duration) *time.Ticker {
 func (l *logStream) collectBatch() {
 	timer := newTicker(batchPublishFrequency)
 	var events []wrappedEvent
-	bytes := 0
+	var eventBuffer []byte
+	var eventBufferTimestamp int64
 	for {
 		select {
-		case <-timer.C:
+		case t := <-timer.C:
+			// If event buffer is older than batch publish frequency flush the event buffer
+			if eventBufferTimestamp > 0 && len(eventBuffer) > 0 {
+				eventBufferAge := t.UnixNano()/int64(time.Millisecond) - eventBufferTimestamp
+				if eventBufferAge > int64(batchPublishFrequency)/int64(time.Millisecond) {
+					events = l.processEvent(events, eventBuffer, eventBufferTimestamp)
+				}
+			}
 			l.publishBatch(events)
 			events = events[:0]
-			bytes = 0
 		case msg, more := <-l.messages:
 			if !more {
+				// Flush event buffer
+				events = l.processEvent(events, eventBuffer, eventBufferTimestamp)
 				l.publishBatch(events)
 				return
 			}
+			if eventBufferTimestamp == 0 {
+				eventBufferTimestamp = msg.Timestamp.UnixNano() / int64(time.Millisecond)
+			}
 			unprocessedLine := msg.Line
-			for len(unprocessedLine) > 0 {
-				// Split line length so it does not exceed the maximum
-				lineBytes := len(unprocessedLine)
-				if lineBytes > maximumBytesPerEvent {
-					lineBytes = maximumBytesPerEvent
+			if l.multilinePattern != nil {
+				if l.multilinePattern.Match(unprocessedLine) {
+					// This is a new log event so flush the current eventBuffer to events
+					events = l.processEvent(events, eventBuffer, eventBufferTimestamp)
+					eventBufferTimestamp = msg.Timestamp.UnixNano() / int64(time.Millisecond)
+					eventBuffer = eventBuffer[:0]
 				}
-				line := unprocessedLine[:lineBytes]
-				unprocessedLine = unprocessedLine[lineBytes:]
-				if (len(events) >= maximumLogEventsPerPut) || (bytes+lineBytes+perEventBytes > maximumBytesPerPut) {
-					// Publish an existing batch if it's already over the maximum number of events or if adding this
-					// event would push it over the maximum number of total bytes.
-					l.publishBatch(events)
-					events = events[:0]
-					bytes = 0
+				eventBuffer = append(eventBuffer, unprocessedLine...)
+				// If we have exceeded max bytes per event flush the event buffer up to max bytes
```

Comment:
I think by splitting only by the `maximumBytesPerEvent`, you lose a nice property of the non-multiline case, which is that a line will only be split if that individual line exceeds `maximumBytesPerEvent`.  With the logic here, you may have a single-character line at the end of a really long multiline event and then a new event with a line where the prefix has been stripped.  Instead, I think it would make sense to have the logic such that we call `processEvent` prior to appending the `unprocessedLine` if appending would put the length of `eventBuffer` over `maximumBytesPerEvent`.  Something like this:

```go
if len(eventBuffer)+len(unprocessedLine) > maximumBytesPerEvent{
	events = l.processEvent(events, eventBuffer, eventBufferTimestamp)
	eventBuffer = [:0]
}

eventBuffer = append(eventBuffer, unprocessedLine...)
```